// @stylin/vite-plugin/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resultCache = void 0;
const pluginutils_1 = require("@rollup/pluginutils");
const parser_js_1 = require("./parser.js");
const node_fs_1 = require("node:fs");
function stringify(o) {
    return JSON.stringify(o);
}
exports.resultCache = new Map();
const filter = (0, pluginutils_1.createFilter)(/\.module\.s?css$/);
function compileCSS(options) {
    const { id, code } = options;
    const rawCSS = (0, node_fs_1.readFileSync)(id, `utf-8`);
    const msa = (0, parser_js_1.parseComments)(rawCSS);
    const mapping = (0, parser_js_1.extractClassNames)(code);
    return {
        mapping,
        msa
    };
}
function stylinLoader() {
    return {
        name: `vite-plugin-stylin`,
        enforce: `post`,
        transform(code, id) {
            if (filter(id)) {
                const { msa, mapping } = compileCSS({ id, code });
                exports.resultCache.set(id, { msa });
                const transformedCode = `import {applyStyle as style, createComponent} from '@stylin/style'
${code.replace(/(export default [\S\s]*;)/, `
const styleComponent = createComponent(${stringify(mapping)})
${msa
                    .map((value) => `export const ${value.componentName} = styleComponent(${stringify(value)})`)
                    .join(`\n`)}
export const applyStyle = style(${stringify(mapping)})(${JSON.stringify(msa)})
$1`)}`;
                return {
                    code: transformedCode,
                    map: { mappings: `` }
                };
            }
        },
        async handleHotUpdate(ctx) {
            const id = ctx.file;
            const code = await ctx.read();
            if (filter(id)) {
                const { msa } = compileCSS({ id, code });
                exports.resultCache.set(id, { msa });
            }
        },
    };
}
exports.default = stylinLoader;
